relay-metrics/src/aggregation.rs

```diff
@@ -1302,17 +1297,38 @@ impl Aggregator {
             return Err(AggregateMetricsErrorKind::InvalidStringLength.into());
         }
 
-        if !protocol::is_valid_mri(&key.metric_name) {
-            relay_log::debug!("invalid metric name {:?}", key.metric_name);
-            relay_log::configure_scope(|scope| {
-                scope.set_extra(
-                    "bucket.project_key",
-                    key.project_key.as_str().to_owned().into(),
-                );
-                scope.set_extra("bucket.metric_name", key.metric_name.into());
-            });
-            return Err(AggregateMetricsErrorKind::InvalidCharacters.into());
-        }
+        key.metric_name = match key.metric_name.parse::<MetricResourceIdentifier>() {
+            Ok(mri) => {
+                if matches!(mri.namespace, MetricNamespace::Unsupported) {
+                    relay_log::debug!("invalid metric namespace {:?}", key.metric_name);
+                    relay_log::configure_scope(|scope| {
+                        scope.set_extra(
+                            "bucket.project_key",
+                            key.project_key.as_str().to_owned().into(),
+                        );
+                        scope.set_extra("bucket.metric_name", key.metric_name.into());
+                    });
+                    return Err(AggregateMetricsErrorKind::InvalidNamespace.into());
+                }
+
+                let mut metric_name = mri.to_string();
+                // do this so cost tracking still works accurately.
+                metric_name.shrink_to_fit();
+                metric_name
+            }
+            Err(_) => {
+                relay_log::debug!("invalid metric name {:?}", key.metric_name);
+                relay_log::configure_scope(|scope| {
+                    scope.set_extra(
+                        "bucket.project_key",
+                        key.project_key.as_str().to_owned().into(),
+                    );
+                    scope.set_extra("bucket.metric_name", key.metric_name.into());
+                });
```

Comment:
_nit_: Could we deduplicate scope configuration with the block above, since both only get called in the error case?

Reply:
i sort of assumed that's too hard wrt borrowing but seems like by brain is still stuck in times before stacked borrows. fixed